### PR TITLE
Add authorization context hash to API logs

### LINKED ISSUE #11449: API log only shows wazuh-wui user for changes from Wazuh App
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.2.x | Wazuh App | Manager/ODFE | Packages/Sources | Centos8 |

Trying to audit changes from Wazuh App in Kibana I noticed that the API log always shows the same 'wazuh-wui' user even when the Kibana user is mapped to a Wazuh user.

Based on [this guide](https://documentation.wazuh.com/current/user-manual/kibana-app/wazuh-rbac.html) I created a new 'test' internal user in Kibana and mapped it for 'all_access' role. Then I mapped it to Wazuh using Roles mapping.

If I logged in Kibana as 'admin' I saw that changes in Wazuh App are recorded in the `/var/ossec/logs/api.log` of the Wazuh manager as 'wazuh-wui'
```
# tail -f /var/ossec/logs/api.log 
2021/12/21 19:19:45 INFO: wazuh-wui X.X.X.X "GET /security/roles" with parameters {"sort": "name"} and body {} done in 0.104s: 200
2021/12/21 19:19:46 INFO: wazuh-wui X.X.X.X "GET /security/rules" with parameters {"sort": "name"} and body {} done in 0.139s: 200
2021/12/21 19:19:46 INFO: wazuh-wui X.X.X.X "GET /security/rules" with parameters {"sort": "name"} and body {} done in 0.048s: 200
```

If I logged in Kibana as 'test' I saw that changes in Wazuh App are recorded in the  `/var/ossec/logs/api.log` of the Wazuh manager also as 'wazuh-wui'
```
# tail -f /var/ossec/logs/api.log 
2021/12/21 19:20:00 INFO: wazuh-wui X.X.X.X "GET /manager/stats/remoted" with parameters {"pretty": ""} and body {} done in 0.018s: 200
2021/12/21 19:20:00 INFO: wazuh-wui X.X.X.X "GET /manager/stats/analysisd" with parameters {"pretty": ""} and body {} done in 0.012s: 200
```

If I authenticated through the Wazuh RestAPI with 'test' user and call and endpoint, it is being recorded as 'test'
```
# TOKEN=$(curl -u test:password -k -X GET "https://X.X.X.X:55000/security/user/authenticate?raw=true")
# curl -k -X GET "https://X.X.X.X:55000/?pretty=true" -H "Authorization: Bearer $TOKEN"
{
   "data": {
      "title": "Wazuh API REST",
      "api_version": "4.2.2",
      "revision": 40216,
      "license_name": "GPL 2.0",
      "license_url": "https://github.com/wazuh/wazuh/blob/4.2/LICENSE",
      "hostname": "example.com",
      "timestamp": "2021-12-21T22:24:01+0000"
   },
   "error": 0
}


# tail -f /var/ossec/logs/api.log 
2021/12/21 19:21:50 INFO: unknown_user X.X.X.X "GET /" with parameters {} and body {} done in 0.001s: 401
2021/12/21 19:22:36 INFO: test X.X.X.X "GET /security/user/authenticate" with parameters {"raw": "true"} and body {} done in 0.322s: 200
```
Is this an expected behavior?
How could I audit changes in Wazuh?

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
| #11449 |

This PR closes #11449. 

In this pull request, I have added a hash that identifies the authorization context used in the run as login. This hash has been added to the API logs and it appears in the run as login requests and requests using tokens generated in run as login requests.

With this change, same Wazuh users using different permissions can be identified properly.

More investigation can be found in the related issue's comments.

### CODE REVIEW

**Comment by mcarmona99** [comment]
## Test results

```
$ pytest api/api/test/ --disable-warnings 
================================================================== test session starts ===================================================================
platform linux -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/manuel/git/wazuh/api
plugins: html-3.1.1, metadata-1.11.0, cov-2.12.0, testinfra-5.0.0, asyncio-0.14.0
collected 273 items                                                                                                                                      

api/api/test/test_alogging.py .............                                                                                                        [  4%]
api/api/test/test_authentication.py ...........                                                                                                    [  8%]
api/api/test/test_configuration.py .........................................                                                                       [ 23%]
api/api/test/test_encoder.py ...                                                                                                                   [ 24%]
api/api/test/test_middlewares.py ..........                                                                                                        [ 28%]
api/api/test/test_uri_parser.py ...                                                                                                                [ 29%]
api/api/test/test_util.py ......................................                                                                                   [ 43%]
api/api/test/test_validator.py ................................................................................................................... [ 85%]
.......................................                                                                                                            [100%]

============================================================ 273 passed, 14 warnings in 1.32s ============================================================
```


```
$ pytest api/api/controllers/test/ --disable-warnings 
================================================================== test session starts ===================================================================
platform linux -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/manuel/git/wazuh/api
plugins: html-3.1.1, metadata-1.11.0, cov-2.12.0, testinfra-5.0.0, asyncio-0.14.0
collected 201 items                                                                                                                                      

api/api/controllers/test/test_active_response_controller.py .                                                                                      [  0%]
api/api/controllers/test/test_agent_controller.py ..........................................                                                       [ 21%]
api/api/controllers/test/test_cdb_list_controller.py ......                                                                                        [ 24%]
api/api/controllers/test/test_ciscat_controller.py .                                                                                               [ 24%]
api/api/controllers/test/test_cluster_controller.py ......................                                                                         [ 35%]
api/api/controllers/test/test_decoder_controller.py .......                                                                                        [ 39%]
api/api/controllers/test/test_default_controller.py .                                                                                              [ 39%]
api/api/controllers/test/test_experimental_controller.py ...............                                                                           [ 47%]
api/api/controllers/test/test_manager_controller.py .................                                                                              [ 55%]
api/api/controllers/test/test_mitre_controller.py .......                                                                                          [ 59%]
api/api/controllers/test/test_overview_controller.py .                                                                                             [ 59%]
api/api/controllers/test/test_rootcheck_controller.py ....                                                                                         [ 61%]
api/api/controllers/test/test_rule_controller.py ........                                                                                          [ 65%]
api/api/controllers/test/test_sca_controller.py ..                                                                                                 [ 66%]
api/api/controllers/test/test_security_controller.py ...................................................                                           [ 92%]
api/api/controllers/test/test_syscheck_controller.py ....                                                                                          [ 94%]
api/api/controllers/test/test_syscollector_controller.py .........                                                                                 [ 98%]
api/api/controllers/test/test_task_controller.py .                                                                                                 [ 99%]
api/api/controllers/test/test_vulnerability_controller.py ..                                                                                       [100%]

============================================================ 201 passed, 9 warnings in 1.97s =============================================================

```

**Review by Kondent — APPROVED** [review]
LGTM

**Comment by mcarmona99** [comment]
## Manual tests after rebase

`api.log`

```
2022/05/12 08:14:54 INFO: Listening on 0.0.0.0:55000..
2022/05/12 08:17:15 INFO: wazuh 172.18.0.1 "GET /security/user/authenticate" with parameters {"raw": "true"} and body {} done in 0.404s: 200
2022/05/12 08:17:20 INFO: wazuh 172.18.0.1 "GET /mitre/mitigations" with parameters {} and body {} done in 0.030s: 200
2022/05/12 08:17:28 INFO: wazuh 172.18.0.1 "GET /agents" with parameters {"select": "name", "search": "worker"} and body {} done in 0.028s: 200
2022/05/12 08:17:32 INFO: wazuh 172.18.0.1 "GET /agents" with parameters {"select": "name", "search": "worker"} and body {} done in 0.021s: 200
2022/05/12 08:17:45 INFO: wazuh (2afb9b83f9314e5d029766197f539792) 172.18.0.1 "POST /security/user/authenticate/run_as" with parameters {} and body {} done in 0.352s: 200
2022/05/12 08:18:11 INFO: wazuh (2afb9b83f9314e5d029766197f539792) 172.18.0.1 "POST /security/user/authenticate/run_as" with parameters {"raw": "true"} and body {} done in 0.359s: 200
2022/05/12 08:18:16 INFO: wazuh (2afb9b83f9314e5d029766197f539792) 172.18.0.1 "GET /agents" with parameters {"select": "name", "search": "worker"} and body {} done in 0.044s: 200
2022/05/12 08:18:22 INFO: wazuh (2afb9b83f9314e5d029766197f539792) 172.18.0.1 "GET /decoders" with parameters {"limit": "1", "select": "filename", "q": "name~agent"} and body {} done in 0.305s: 200
2022/05/12 08:18:27 INFO: wazuh (2afb9b83f9314e5d029766197f539792) 172.18.0.1 "GET /decoders" with parameters {} and body {} done in 0.292s: 200
```

`api.json`:

```
{"timestamp": "2022/05/12 08:14:54", "levelname": "INFO", "data": {"type": "informative", "payload": "Listening on 0.0.0.0:55000.."}}
{"timestamp": "2022/05/12 08:17:15", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /security/user/authenticate", "parameters": {"raw": "true"}, "body": {}, "time": "0.404s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:17:20", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /mitre/mitigations", "parameters": {}, "body": {}, "time": "0.030s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:17:28", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /agents", "parameters": {"select": "name", "search": "worker"}, "body": {}, "time": "0.028s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:17:32", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /agents", "parameters": {"select": "name", "search": "worker"}, "body": {}, "time": "0.021s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:17:45", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "hash_auth_context": "2afb9b83f9314e5d029766197f539792", "ip": "172.18.0.1", "http_method": "POST", "uri": "POST /security/user/authenticate/run_as", "parameters": {}, "body": {}, "time": "0.352s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:18:11", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "hash_auth_context": "2afb9b83f9314e5d029766197f539792", "ip": "172.18.0.1", "http_method": "POST", "uri": "POST /security/user/authenticate/run_as", "parameters": {"raw": "true"}, "body": {}, "time": "0.359s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:18:16", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "hash_auth_context": "2afb9b83f9314e5d029766197f539792", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /agents", "parameters": {"select": "name", "search": "worker"}, "body": {}, "time": "0.044s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:18:22", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "hash_auth_context": "2afb9b83f9314e5d029766197f539792", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /decoders", "parameters": {"limit": "1", "select": "filename", "q": "name~agent"}, "body": {}, "time": "0.305s", "status_code": 200}}}
{"timestamp": "2022/05/12 08:18:27", "levelname": "INFO", "data": {"type": "request", "payload": {"user": "wazuh", "hash_auth_context": "2afb9b83f9314e5d029766197f539792", "ip": "172.18.0.1", "http_method": "GET", "uri": "GET /decoders", "parameters": {}, "body": {}, "time": "0.292s", "status_code": 200}}}
```

File: api/api/alogging.py [hunk 1]
```diff
@@ -46,15 +49,17 @@ def custom_logging(self, user, remote, method, path, query, body, time, status):
             Required time to compute the request.
         status : int
             Status code of the request.
+        hash_auth_context : str, optional
+            Hash representing the authorization context. Default: ''
         """
 
         self.logger.info(f'{user} '
+                         f'{f"({hash_auth_context}) " if hash_auth_context else ""}'
```
Comment (vicferpoy) [inline api/api/alogging.py:57]: Remove redundant f-strings:

```python
>>> a = False
>>> 
>>> f"({a})" if a else ""
>>> ''

>>> a = "hash"
>>> 
>>> f"({a})" if a else ""
'(hash)'
```

File: api/api/authentication.py [hunk 2]
```diff
@@ -161,15 +163,16 @@ def generate_token(user_id=None, data=None, run_as=False):
...
+                  "iss": JWT_ISSUER,
+                  "aud": "Wazuh API REST",
+                  "nbf": timestamp,
+                  "exp": timestamp + result['auth_token_exp_timeout'],
+                  "sub": str(user_id),
+                  "run_as": auth_context is not None,
+                  "rbac_roles": data['roles'],
+                  "rbac_mode": result['rbac_mode']
+              } | ({"hash_auth_context": hashlib.blake2b(json.dumps(auth_context).encode(), digest_size=16).hexdigest()}
+                   if auth_context is not None else {})
```
Comment (vicferpoy) [inline api/api/authentication.py:175]: What happened to the indentation?

**Review by vicferpoy — CHANGES_REQUESTED** [review]

File: api/api/alogging.py (hunk 1 above)
Comment (mcarmona99) [reply api/api/alogging.py:57]: The f-strings are not redundant but the truth is that the readability could be improved. I have included some changes to do it.

File: api/api/authentication.py (hunk 2 above)
Comment (mcarmona99) [reply api/api/authentication.py:175]: It was updated in this PR.

**Review by vicferpoy — APPROVED** [review]
LGTM

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**